custom/icds_reports/tasks.py

```diff
@@ -74,14 +75,31 @@ def move_ucr_data_into_aggregation_tables(date=None, intervals=2):
             _create_aggregate_functions(cursor)
             _update_aggregate_locations_tables(cursor)
 
-        aggregation_tasks = []
+        tasks = []
 
         for monthly_date in monthly_dates:
             calculation_date = monthly_date.strftime('%Y-%m-%d')
-            aggregation_tasks.append(UCRAggregationTask('monthly', calculation_date))
-
-        aggregation_tasks.append(UCRAggregationTask('daily', date.strftime('%Y-%m-%d')))
-        aggregate_tables.delay(aggregation_tasks[0], aggregation_tasks[1:])
+            tasks.extend([
+                group(
+                    icds_aggregation_task.si(date=calculation_date, func=_update_months_table),
+                    icds_aggregation_task.si(date=calculation_date, func=_aggregate_cf_forms),
+                    icds_aggregation_task.si(date=calculation_date, func=_aggregate_thr_forms),
+                ),
+                group(
+                    icds_aggregation_task.si(date=calculation_date, func=_child_health_monthly_table),
+                    icds_aggregation_task.si(date=calculation_date, func=_ccs_record_monthly_table),
+                    icds_aggregation_task.si(date=calculation_date, func=_daily_attendance_table),
+                ),
+                group(
+                    icds_aggregation_task.si(date=calculation_date, func=_agg_child_health_table),
+                    icds_aggregation_task.si(date=calculation_date, func=_agg_ccs_record_table),
+                ),
+                icds_aggregation_task.si(date=calculation_date, func=_agg_awc_table),
+            ])
+
+        tasks.append(icds_aggregation_task.si(date=date.strftime('%Y-%m-%d'), func=aggregate_awc_daily))
+        tasks.append(email_dashboad_team.si(aggregation_date=date.strftime('%Y-%m-%d')))
```

Comment:
curious how you chose these groupings / order. guessing maybe requires a complex amount of knowledge of the underlying functions? just wondering how this will get maintained over time or if that is a concern from a knowledge/documentation perspective.

Reply:
The workflow for the aggregation is documented [here](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/docs/current_state_aggregation.png) (generated with plantuml from [this](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/docs/current_state_aggregation.txt)) with the ideal workflow [here](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/docs/ideal_aggregation_workflow.png) (from my original spec)

These groupings were chosen by figuring out what tables had no dependencies (first group), then all the tables that could be built with those tables (second group) and so on until all were built.

Whenever a new table is added it would be added to this workflow in the earliest grouping as possible.

>  I'm not too familiar with celery parallellization

FWIW I don't think we actually use this anywhere yet, so no one may be familiar with it :)

Reply:
nice, this is awesome. so basically anything can be run if the all the boxes pointing into it have run, yeah?

Reply:
Yeah. that's the idea at least. So far that's the best way I've found to document the workflow, though I imagine using airflow like the warehouse might be better if we start using that

Reply:
That's pretty neat.  I'm guessing the run-time of these tasks dwarfs any overhead associated with spinning up a bunch of tasks rather than one big one?